### ISSUE TITLE
core: improve experience around stack-level tagging

### ISSUE THREAD

**Issue by eladb** [aws/aws-cdk#4134]
The issue is that you applying the tag on the entire stack (`self` as the first parameter to `Tag.add`), which means that the CLI attempts to apply these tags to the stack, which can only be done when it deploys through CloudFormation. Since `{ "Ref": "AWS::StackName" }` is only resolved *during* deployment, the result is this error.

We won't be able to support deploy-time values in stack tags, but we should definitely improve this experience:

1. We should emit an error if you try to apply a tag that uses tokens to the Stack itself.
2. We should give you a way to opt-out of stack tags

In the meantime, you can do two things:

1. Use `stack.stackName` (which resolves to the concrete stack name) instead of `Aws.STACK_NAME`.
2. Apply the tag to a specific resource (i.e. the bucket: `Tag.add(r3, 'Instance', ...)`
3. If your goal is to apply the same tag to all resources in your stack, you can wrap them in a construct and apply the tag to the construct:

```python
group = Construct(self, 'Group')

r3 = s3.Bucket(group, 'Bucket')
Tag.add(group, 'Instance', Aws.STACK_NAME)
```

Let me know if this helps.

_Originally posted by @eladb in https://github.com/aws/aws-cdk/issues/4106#issuecomment-532626191_

**Comment by brettswift** [comment]
I'm not sure if this belongs here - It's related to my experience with a tagging abstraction attempt. 

A common scenario we have is tagging standards.    I wanted to create something that enforced the conventions and simplified the usage.  
I first tried a static helper method that took a construct and a bunch of tag values, but was getting `method is undefined` - seemed to be a weird typescript error.  

So I created a construct to try that. 

The usage of my construct is as follows, attempting to add tags at the App level. 
```
const app = new cdk.App();

new ExampleApi(app, 'ConfigAutomationExampleApiStack');

new CorporateTags(app, 'CorporateTags', {
    Environment: 'bswift',
    OwnerContact: 'me@email.com',
    SolutionName: 'ExampleApi',
}as CorporateTagProps)

cdk.Tag.add(app, 'TEST-SolutionName', 'SampleApi')
```

The class looks something like this: 
```
~
    constructor(scope: cdk.Construct, id: string, props: CorporateTagProps) {
        super(scope, id);
        
        cdk.Tag.add(scope, this.CE_ENVIRONMENT, props.Environment)
        cdk.Tag.add(scope, this.CE_SOLUTIONNAME, props.SolutionName)
```

When deployed I see the `TEST-SolutionName` but not the other two.   I thought the visitor pattern used in the Tag class would result in the injected scope being tagged.  However that is not happening. 

Is there an approach that I could do differently here?  Or is this a bug?  I'd like something more than just exporting constants for apps to use.
